site/content/entry/citra-progress-report-2018-Q1.md

```diff
@@ -0,0 +1,230 @@
+code into smaller parts, fixing and testing them.
+
+Thanks to [degasus](https://github.com/degasus) from the Dolphin community, who imparted his valuable
+knowledge and understanding of GPUs and OpenGL, we were able to optimize the _streaming storage buffer_
+support added by [phantom](https://github.com/phanto-m).This was done entirely by [degasus](https://github.com/degasus)
+himself. ([here](https://github.com/citra-emu/citra/pull/3504),[here](https://github.com/citra-emu/citra/pull/3666)
+and [here](https://github.com/citra-emu/citra/pull/3711)). The old way to upload the data (both vertex
+and constants) was a simple `glBufferData` call. Keep in mind, however, that the GPU has a huge execution
+latency so you can't just copy the new content over the existing buffer. The driver has to choose
+between different ways to deal with this issue, with different drawbacks:
+
+1. Wait for the GPU to finish and copy to the destination buffer. Usually this method is the fastest
+for big uploads, fine on common gaming, but terrible for emulators.
+2. Allocate a new buffer and tell the GPU which new buffer shall be used now.You also need to care
+about freeing the old buffer once it isn't in use any more. This tends to a huge CPU overhead and
+isn't available on partial buffer updates such as Lighting LUTs.
+3. Copy the data to an internal staging buffer and ask the GPU to copy it again to the destination
+when it is ready. It uses twice the memory bandwidth, and switching between rendering and copy also takes a
+while for the GPU. This proves to be a significant bottleneck, if you do this per draw call.
+
+In the end, all of these are bad in terms of performance. It is a simple upload call, but it either
+requires many copies and context switches, or it stalls the GPU (_More on that in [here](https://de.slideshare.net/CassEveritt/approaching-zero-driver-overhead)_). Our buffer streaming
+way is a bit different. Now, we allocate a big staging buffer ourselves, copy the data to the current
+position within the buffer, and tell the GPU to directly access the data from there. The driver has
+no scope of interpreting what we ask it to do.
+
+_This has a downside as well_. The GPU now accesses the main memory over PCIe. So if we're going to
+reuse the same data in the next frame, it has to go through the PCIe bottleneck again. But, we never
+use the same data again and hence gain a big performance boost.
+
+[wwylele](https://github.com/wwylele) tweaked the configuration UI and renamed some parts of it.
+However, this caused some confusion about the configuration window and GPU features.
+
+### What's changed in the UI ?
+
+With the initial release of GLvtx, we've had support for running shaders on the host GPU . So, we had
+a dropdown menu called **Shader Emulation** to select either CPU or GPU. The option has now been renamed
+as **Enable Hardware Shader**. When enabled, it makes use of the host GPU, and when disabled, it uses
+the CPU.
```

Comment:
The reason why @wwylele changed the name is that he thinks the ```CPU``` and ```GPU``` names are not accurate. The "CPU" implementation uses the GPU as well.
You should say ```it mainly uses the CPU```

Reply:
When the `Enable Hardware Shader` option is enabled, shaders run mostly on host GPU (except in rare cases). when disabled shaders run on CPU alone. (as confirmed by wwylele)